src/Microsoft.OData.Core/Json/ODataJsonReaderCoreUtils.cs

```diff
@@ -220,5 +324,125 @@ private static IEnumerable<object> ReadArrayValue(IJsonReader jsonReader, ODataI
+                await jsonReader.ReadNextAsync()
+                    .ConfigureAwait(false);
+            }
+
+            while (jsonReader.NodeType != JsonNodeType.EndObject)
+            {
+                // read the property name
+                string propertyName = await jsonReader.ReadPropertyNameAsync()
+                    .ConfigureAwait(false);
+
+                // read the property value
+                object propertyValue;
+                switch (jsonReader.NodeType)
+                {
+                    case JsonNodeType.PrimitiveValue:
+                        propertyValue = await jsonReader.ReadPrimitiveValueAsync()
+                            .ConfigureAwait(false);
+                        break;
+                    case JsonNodeType.StartArray:
+                        propertyValue = await ReadArrayValueAsync(jsonReader, inputContext, recursionDepth)
+                            .ConfigureAwait(false);
+                        break;
+                    case JsonNodeType.StartObject:
+                        propertyValue = await ReadObjectValueAsync(jsonReader, /*insideJsonObjectValue*/ false, inputContext, recursionDepth)
+                            .ConfigureAwait(false);
+                        break;
+                    default:
+                        Debug.Assert(false, "We should never reach here - There should be matching end element");
+                        return null;
+                }
+
+                jsonValue.Add(ODataAnnotationNames.RemoveAnnotationPrefix(propertyName), propertyValue);
+            }
+
+            await jsonReader.ReadEndObjectAsync()
+                .ConfigureAwait(false);
+
+            return jsonValue;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the json array from the reader.
+        /// </summary>
+        /// <param name="jsonReader">JsonReader instance.</param>
+        /// <param name="inputContext">The input context with all the settings.</param>
+        /// <param name="recursionDepth">The recursion depth to start with.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+		/// The value of the TResult parameter contains a lit of JSON objects.</returns>
+        /// <returns>a list of json objects.</returns>
+        private static async Task<IEnumerable<object>> ReadArrayValueAsync(IJsonReaderAsync jsonReader, ODataInputContext inputContext, int recursionDepth)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+            Debug.Assert(jsonReader.NodeType == JsonNodeType.StartArray, "jsonReader.NodeType == JsonNodeType.StartArray");
+            Debug.Assert(inputContext != null, "inputContext != null");
+
+            ValidationUtils.IncreaseAndValidateRecursionDepth(ref recursionDepth, inputContext.MessageReaderSettings.MessageQuotas.MaxNestingDepth);
+
+            List<object> array = new List<object>();

```

Comment:
nit: calling a non-array item an array

```suggestion
            List<object> items= new List<object>();

```